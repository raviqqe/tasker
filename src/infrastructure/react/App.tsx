import { PulseLoader } from "react-spinners";
import { useAsync } from "react-use";
import React from "react";
import styled from "styled-components";
import { observer } from "mobx-react";
import { AuthenticationStore } from "../mobx/authentication-store";
import { DocumentsStore } from "../mobx/documents-store";
import { IProps as ILandingProps, Landing } from "./Landing";
import { Home, IProps as IHomeProps } from "./Home";

const LoaderContainer = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  height: 100vh;
  width: 100vw;
`;

interface IProps extends Omit<IHomeProps, "documents">, ILandingProps {
  authenticationStore: AuthenticationStore;
  documentsStore: DocumentsStore;
  initialize: () => Promise<void>;
}

export const App = observer(
  ({
    authenticationStore: { signedIn },
    documentsStore: { documents },
    initialize,
    repositoryURL,
    signIn,
    signOut,
    ...props
  }: IProps) => {
    useAsync(initialize, []);

    return signedIn === null ? (
      <LoaderContainer>
        <PulseLoader color="white" />
      </LoaderContainer>
    ) : signedIn ? (
      <Home {...props} documents={documents} signOut={signOut} />
    ) : (
      <Landing repositoryURL={repositoryURL} signIn={signIn} />
    );
  }
);
