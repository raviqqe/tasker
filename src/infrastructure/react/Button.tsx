import styled from "styled-components";
import { boxShadow } from "./style";
import { red, grey, white } from "./style/colors";

export const Button = styled.button<{ secondary?: boolean }>`
  ${boxShadow};
  background: ${({ secondary }) => (secondary ? grey : red)};
  display: flex;
  justify-content: center;
  align-items: center;
  font-family: inherit;
  color: ${white};
  border: none;
  cursor: pointer;
  flex-shrink: 0;
`;
