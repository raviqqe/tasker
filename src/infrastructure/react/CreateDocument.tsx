import { MdAdd } from "react-icons/md";
import React, { useState } from "react";
import styled from "styled-components";
import { CircleButton } from "./CircleButton";
import { InsertFilesFunction } from "./utilities";
import { MarkdownTextArea } from "./MarkdownTextArea";

const Container = styled.div`
  display: flex;
  align-items: center;
  padding: 1em;
`;

const StyledMarkdownTextArea = styled(MarkdownTextArea)`
  margin-right: -1em;
  max-height: 80vh;
`;

interface IProps {
  createDocument: (text: string) => Promise<void>;
  insertFiles: InsertFilesFunction;
}

export const CreateDocument = ({ createDocument, insertFiles }: IProps) => {
  const [text, setText] = useState("");
  const onSubmit = async (): Promise<void> => {
    setText("");
    await createDocument(text);
  };

  return (
    <Container>
      <StyledMarkdownTextArea
        insertFiles={insertFiles}
        onSubmit={onSubmit}
        setText={setText}
        text={text}
      />
      <CircleButton aria-label="Create" onClick={onSubmit}>
        <MdAdd />
      </CircleButton>
    </Container>
  );
};
