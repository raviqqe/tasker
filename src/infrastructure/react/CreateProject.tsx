import { MdAdd } from "react-icons/md";
import React from "react";
import { CircleButton } from "./CircleButton";

export interface IProps {
  createProject: (name: string) => Promise<void>;
}

export const CreateProject = ({ createProject, ...restProps }: IProps) => (
  <CircleButton
    aria-label="Create Project"
    onClick={async () => {
      const name = window.prompt("Project name?");

      if (!name) {
        return;
      }

      await createProject(name);
    }}
    {...restProps}
  >
    <MdAdd />
  </CircleButton>
);
