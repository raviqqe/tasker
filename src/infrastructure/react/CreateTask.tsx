import { MdAdd } from "react-icons/md";
import React from "react";
import { CircleButton } from "./CircleButton";

interface IProps {
  createTask: (name: string) => Promise<void>;
}

export const CreateTask = ({ createTask }: IProps) => {
  return (
    <CircleButton
      aria-label="Create"
      onClick={async () => {
        const name = window.prompt("Task name?");

        if (!name) {
          return;
        }

        await createTask(name);
      }}
    >
      <MdAdd />
    </CircleButton>
  );
};
