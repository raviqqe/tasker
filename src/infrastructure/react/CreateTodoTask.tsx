import { MdAdd } from "react-icons/md";
import { CircleButton } from "./CircleButton";

export interface IProps {
  createTodoTask: (name: string) => Promise<void>;
}

export const CreateTodoTask = ({ createTodoTask, ...restProps }: IProps): JSX.Element => (
  <CircleButton
    aria-label="Create"
    onClick={async () => {
      const name = window.prompt("Task name?");

      if (!name) {
        return;
      }

      await createTodoTask(name);
    }}
    {...restProps}
  >
    <MdAdd />
  </CircleButton>
);
