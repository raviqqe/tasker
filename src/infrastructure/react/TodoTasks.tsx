import { PulseLoader } from "react-spinners";
import React from "react";
import { SortableContainer, SortableElement } from "react-sortable-hoc";
import styled from "styled-components";
import { ITask } from "../../domain/task";
import { Task } from "./Task";
import { buttonMargin } from "./style";

const Container = styled.div`
  display: flex;
  flex-direction: column;
  overflow: auto;
`;

const LoaderContainer = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
`;

const Tasks = SortableContainer(styled.div`
  display: flex;
  flex-direction: column;
  padding: 1em 0.5em ${buttonMargin};

  > * {
    margin: 0.5em;
  }
`);

const SortableTask = SortableElement(Task);

export interface IProps {
  completeTodoTask: (task: ITask) => Promise<void>;
  reorderTodoTasks: (taskIDs: string[]) => Promise<void>;
  todoTasks: ITask[] | null;
  updateTodoTask: (task: ITask) => Promise<void>;
}

export const TodoTasks = ({
  completeTodoTask,
  reorderTodoTasks,
  todoTasks,
  updateTodoTask
}: IProps) =>
  todoTasks ? (
    <Container>
      <Tasks
        distance={4}
        onSortEnd={async ({ oldIndex, newIndex }) => {
          const taskIDs = todoTasks.map(task => task.id);
          taskIDs.splice(newIndex, 0, taskIDs.splice(oldIndex, 1)[0]);
          await reorderTodoTasks(taskIDs);
        }}
        useDragHandle={true}
      >
        {todoTasks.map((task: ITask, index: number) => (
          <SortableTask
            completeTask={completeTodoTask}
            dragHandleEnabled={true}
            index={index}
            key={task.id}
            task={task}
            updateTask={updateTodoTask}
          />
        ))}
      </Tasks>
    </Container>
  ) : (
    <LoaderContainer>
      <PulseLoader color="white" />
    </LoaderContainer>
  );
