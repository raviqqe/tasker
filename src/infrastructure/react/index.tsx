import { render } from "react-dom";
import React from "react";
import { ApplicationInitializer } from "../../application/application-initializer";
import { DocumentCreator } from "../../application/document-creator";
import { DocumentLister } from "../../application/document-lister";
import { DocumentUpdater } from "../../application/document-updater";
import { IDocument } from "../../domain/document";
import { SignInManager } from "../../application/sign-in-manager";
import { SignOutManager } from "../../application/sign-out-manager";
import { TextFileInserter } from "../../application/text-file-inserter";
import { AuthenticationStore } from "../mobx/authentication-store";
import { DocumentsStore } from "../mobx/documents-store";
import { GlobalStyle } from "./style";
import { App } from "./App";

export class ReactRenderer {
  constructor(
    private readonly applicationInitializer: ApplicationInitializer,
    private readonly documentCreator: DocumentCreator,
    private readonly documentLister: DocumentLister,
    private readonly documentUpdater: DocumentUpdater,
    private readonly signInManager: SignInManager,
    private readonly signOutManager: SignOutManager,
    private readonly textFileInserter: TextFileInserter,
    private readonly authenticationStore: AuthenticationStore,
    private readonly documentsStore: DocumentsStore,
    private readonly repositoryURL: string
  ) {}

  public render(element: HTMLElement): void {
    render(
      <>
        <App
          authenticationStore={this.authenticationStore}
          documentsStore={this.documentsStore}
          createDocument={(text: string) => this.documentCreator.create(text)}
          initialize={() => this.applicationInitializer.initialize()}
          insertFiles={(
            text: string,
            position: number,
            files: File[]
          ): Promise<string> =>
            this.textFileInserter.insert(text, position, files)
          }
          listDocuments={() => this.documentLister.list()}
          listMoreDocuments={() => this.documentLister.listMore()}
          repositoryURL={this.repositoryURL}
          signIn={() => this.signInManager.signIn()}
          signOut={() => this.signOutManager.signOut()}
          updateDocument={(document: IDocument, text: string) =>
            this.documentUpdater.update(document, text)
          }
        />
        <GlobalStyle />
      </>,
      element
    );
  }
}
