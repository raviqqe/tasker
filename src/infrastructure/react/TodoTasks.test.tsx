import { render } from "@testing-library/react";
import { TodoTasks } from "./TodoTasks";
import { it, expect } from "vitest";

it("renders", () => {
  expect(
    render(
      <TodoTasks
        completeTodoTask={async () => {}}
        reorderTodoTasks={async () => {}}
        todoTasks={[{ id: "id", name: "name" }]}
        updateTodoTask={async () => {}}
      />
    ).container.firstChild
  ).toMatchSnapshot();
});

it("renders with no tasks", () => {
  expect(
    render(
      <TodoTasks
        completeTodoTask={async () => {}}
        reorderTodoTasks={async () => {}}
        todoTasks={[]}
        updateTodoTask={async () => {}}
      />
    ).container.firstChild
  ).toMatchSnapshot();
});

it("renders with tasks not loaded yet", () => {
  expect(
    render(
      <TodoTasks
        completeTodoTask={async () => {}}
        reorderTodoTasks={async () => {}}
        todoTasks={null}
        updateTodoTask={async () => {}}
      />
    ).container.firstChild
  ).toMatchSnapshot();
});
