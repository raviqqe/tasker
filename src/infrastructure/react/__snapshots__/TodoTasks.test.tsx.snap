// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders 1`] = `
<div
  class="sc-jSMdHm hBAMbu"
>
  <div
    class="sc-iBkiHk gHtwft"
  >
    <div
      aria-describedby="DndDescribedBy-0"
      aria-disabled="false"
      aria-roledescription="sortable"
      class="sc-gsnScm fWpygO"
      role="button"
      tabindex="0"
    >
      <div
        class="sc-dkzCjb Mhifw"
      >
        name
      </div>
      <div
        class="sc-hKMtDE gnRsPn"
      >
        <div
          aria-label="Done"
          class="sc-bczSft jPkYYc"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M9 16.17L4.83 12l-1.42 1.41L9 19 21 7l-1.41-1.41z"
            />
          </svg>
        </div>
        <div
          aria-label="Edit"
          class="sc-bczSft jPkYYc"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04a.996.996 0 000-1.41l-2.34-2.34a.996.996 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
            />
          </svg>
        </div>
        <div
          class="sc-bczSft jPkYYc sc-eCYdKt drCLIA"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M20 9H4v2h16V9zM4 15h16v-2H4v2z"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders with no tasks 1`] = `
<div
  class="sc-jSMdHm hBAMbu"
>
  <div
    class="sc-iBkiHk gHtwft"
  />
</div>
`;

exports[`renders with tasks not loaded yet 1`] = `
<div
  class="sc-gKXNOb insiZa"
>
  <span
    style="display: inherit;"
  >
    <span
      style="background-color: white; width: 15px; height: 15px; margin: 2px; border-radius: 100%; display: inline-block; animation: react-spinners-PulseLoader-pulse 0.75s 0.12s infinite cubic-bezier(0.2, 0.68, 0.18, 1.08); animation-fill-mode: both;"
    />
    <span
      style="background-color: white; width: 15px; height: 15px; margin: 2px; border-radius: 100%; display: inline-block; animation: react-spinners-PulseLoader-pulse 0.75s 0.24s infinite cubic-bezier(0.2, 0.68, 0.18, 1.08); animation-fill-mode: both;"
    />
    <span
      style="background-color: white; width: 15px; height: 15px; margin: 2px; border-radius: 100%; display: inline-block; animation: react-spinners-PulseLoader-pulse 0.75s 0.36s infinite cubic-bezier(0.2, 0.68, 0.18, 1.08); animation-fill-mode: both;"
    />
  </span>
</div>
`;
